### PR TITLE
Add more requant methods

### PR DESCRIPTION
**Pull request by vikramdattu**
1. Added more versions of requant
2. Fixed some warnings for ESP32 build

### CODE REVIEW

File: tensorflow/lite/kernels/internal/reference/requantize.h [hunk 1]
```diff
@@ -45,6 +45,7 @@ inline void Requantize(const input_type* input_data, int32_t size,
       for (int i = 0; i < size; ++i) {
         output_data[i] = input_data[i] ^ 0x80;
       }
+      return;
```
Comment (advaitjain) [inline tensorflow/lite/kernels/internal/reference/requantize.h:48]: this isn't really doing anything. I'd say remove.

**Review by advaitjain — COMMENTED** [review]
Since the uint8 is deprecated (https://www.tensorflow.org/lite/performance/quantization_spec), we would like to avoid adding support unless there is a strong reason for this.

If you can provide more context (via a github issue) on why you need uint8 requantization then that would help figure out the right path forward.

And it would be great if you could separate the ESP32 build warning fixes into a separate pull request, that would be great.

File: tensorflow/lite/kernels/internal/reference/requantize.h (hunk 1 above)
Comment (advaitjain) [reply tensorflow/lite/kernels/internal/reference/requantize.h:48]: actually, I take that back. This return is indeed changing the behavior.

The change looks very reasonable to me but let's move this to its own pull request too. Feel free to tag me on the PRs and I can review them quickly.

**Comment by vikramdattu** [comment]
> Since the uint8 is deprecated (https://www.tensorflow.org/lite/performance/quantization_spec), we would like to avoid adding support unless there is a strong reason for this.
> 
> If you can provide more context (via a github issue) on why you need uint8 requantization then that would help figure out the right path forward.
> 
> And it would be great if you could separate the ESP32 build warning fixes into a separate pull request, that would be great.


@advaitjain I have moved both commits to separate PRs.

About uint8 deprecation, I think I have totally missed this. And since my model used uint8 input/output, it added uint8->int8 and int8->uint8 quantization, which obviously is not present in current implementation.

I am moving this PR to draft.

**Comment by vikramdattu** [comment]
@advaitjain I think we still need requant methods?

Raw images are uint8_t in nature. Pixel value ranging from [0, 255]!
Don't we need a requant (uint8-->int8) at the very first step?

Getting confused here!

I use following lines in my model conversion code:
```
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.uint8
model_tflite = converter.convert()
```

The reason is simple. Images are in fact uint8 pixels and it will need uint8-->int8 conversion before been feed to other OPs.

**Comment by advaitjain** [comment]
You're right, what is deprecated is asymmetric quantization (https://github.com/tensorflow/tensorflow/issues/44912). Accepting uint8 as model inputs, for example via images as you describe should still be supported.

Your particular use-case makes sense, Let's add for only this specific case of input == uint8, output == int8 and  document the reason.

What I want to avoid is general uint8 support (e.g. taking an int8 input and making it uint8 which then means that the ops following will need to support asymmetric quantization).

**Comment by vikramdattu** [comment]
@advaitjain thanks.
The other version is also needed IMHO.

The output if has a range value should be in [0, 255] i.e., in uint8 format and not int8.
Having int8-->uint8 version helps that.

Of course interpreting int8 output is option but having [int8->uint8] version seems much cleaner.

**Comment by advaitjain** [comment]
If it works for you then let's have the output type for the model be `int8` instead of `uint8` and see how far we can go.

**Comment by gbaned** [comment]
@vikramdattu  This PR is in draft, any update on this? Please. Thanks!

**Comment by vikramdattu** [comment]
> @vikramdattu This PR is in draft, any update on this? Please. Thanks!

All of the changes are either merged via separate PRs or are to be merged. (Except int8_t->uint8_t requant, which this PR contains. And as per discussions we want to avoid it).

@gbaned this MR will be closed. Will do it once a pending PR https://github.com/tensorflow/tensorflow/pull/45193 is merged.

**Pull request closed without merging**